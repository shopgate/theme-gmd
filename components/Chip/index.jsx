/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '@shopgate/pwa-common/components/Button';
import CrossIcon from 'Components/icons/CrossIcon';
import colors from 'Styles/colors';
import styles from './style';

/**
 * The Chip component.
 */
class Chip extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    invert: PropTypes.bool,
    onClick: PropTypes.func,
    onRemove: PropTypes.func,
    removable: PropTypes.bool,
  };

  static defaultProps = {
    invert: false,
    onClick: () => {},
    onRemove: () => {},
    removable: true,
  };

  /**
   * When the component updates make sure that inline styles are removed.
   * This is required because even when re-rendering the component, react doesn't
   * check the inline stylings.
   * We therefore explicitly force to remove the styling after rendering.
   * ChipsLayout will add custom styles and expects the component to not have any styling
   * after re-rendering.
   */
  componentDidUpdate() {
    this.elementRef.removeAttribute('style');
  }

  /**
   * Returns a class name for the wrapper.
   * @returns {string}
   */
  get wrapperStyle() {
    return styles.chip(this.props.removable, this.props.invert);
  }

  /**
   * Returns the chip's font color.
   * @returns {string}
   */
  get crossFontColor() {
    if (this.props.invert) {
      return colors.light;
    }

    return colors.accent;
  }

  /**
   * Renders the remove icon of the chip if it can be removable.
   * @returns {JSX|null}
   */
  renderRemoveIcon() {
    if (!this.props.removable) {
      return null;
    }

    return (
      <Button className={styles.removeButton} onClick={this.props.onRemove}>
        <CrossIcon size={16} color={this.crossFontColor} />
      </Button>
    );
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    return (
      <div
        ref={(element) => { this.elementRef = element; }}
        className={this.wrapperStyle}
      >
        {this.renderRemoveIcon()}
        <Button className={styles.name} onClick={this.props.onClick}>
          {this.props.children}
        </Button>
      </div>
    );
  }
}

export default Chip;
