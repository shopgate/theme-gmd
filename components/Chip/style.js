/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';
import colors from 'Styles/colors';
import variables from 'Styles/variables';

/**
 * Gets a basic style object for the chip layout.
 * @param {boolean} hasRemoveButton Whether this chip has a remove button.
 * @returns {Object} The style object.
 */
const chipBase = (hasRemoveButton = true) => ({
  display: 'flex',
  alignItems: 'center',
  borderRadius: 26,
  outline: 0,
  height: 26,
  paddingRight: variables.gap.small,
  paddingLeft: variables.gap.small * (hasRemoveButton ? 0.5 : 1),
  marginRight: 5,
  marginTop: 5,
  marginBottom: 5,
  minWidth: 0,
});

/**
 * Gets a style class for the chip layout.
 * @param {boolean} hasRemoveButton Whether this chip has a remove button.
 * @param {boolean} inverted Whether the colors of the chip are inverted.
 * @returns {string} The style class name.
 */
const chip = (hasRemoveButton = true, inverted = false) => css({
  ...chipBase(hasRemoveButton),
  backgroundColor: (inverted ? colors.accent : colors.accentContrast),
  color: (inverted ? colors.accentContrast : colors.accent),
}).toString();

const removeButton = css({
  flexShrink: 0,
  margin: 0,
  padding: 0,
}).toString();

const name = css({
  paddingLeft: (variables.gap.small * 0.5),
  paddingRight: (variables.gap.small * 0.5),
  paddingTop: 3,
  paddingBottom: 3,
  fontSize: 12,
  fontWeight: 500,
  textOverflow: 'ellipsis',
  maxWidth: '100%',
  whiteSpace: 'nowrap',
  overflow: 'hidden',
  display: 'block',
  lineHeight: '1',
  color: 'inherit',
}).toString();

export default {
  chip,
  removeButton,
  name,
};
