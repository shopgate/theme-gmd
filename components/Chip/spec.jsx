/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Chip from 'Components/Chip';

describe('<Chip />', () => {
  it('should render a tag', () => {
    const wrapper = mount(<Chip>text</Chip>);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('button').length).toEqual(2);
    expect(wrapper.find('button').at(1).text()).toEqual('text');
  });
});

describe('<Chip />', () => {
  it('should render a without removable icon', () => {
    const wrapper = mount(<Chip removable={false}>text</Chip>);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('button').length).toEqual(1);
    expect(wrapper.find('button').at(0).text()).toEqual('text');
  });
});
