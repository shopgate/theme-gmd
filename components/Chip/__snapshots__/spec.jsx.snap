// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chip /> should render a tag 1`] = `
<Chip
  invert={false}
  onClick={[Function]}
  onRemove={[Function]}
  removable={true}
>
  <div
    className="css-h0zu4c"
  >
    <Button
      className="css-i7sqdh"
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="css-i7sqdh css-1rn65c0"
        data-test-id="Button"
        disabled={false}
        onClick={[Function]}
      >
        <Cross
          color="#5ccee3"
          size={16}
        >
          <Icon
            className=""
            color="#5ccee3"
            content="<path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"/>"
            size={16}
            viewBox="0 0 24 24"
          >
            <svg
              className="css-yys9hb "
              dangerouslySetInnerHTML={
                Object {
                  "__html": "<path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"/>",
                }
              }
              style={
                Object {
                  "fill": "#5ccee3",
                  "fontSize": 16,
                }
              }
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </Icon>
        </Cross>
      </button>
    </Button>
    <Button
      className="css-1al9zi1"
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="css-1al9zi1 css-1rn65c0"
        data-test-id="Button"
        disabled={false}
        onClick={[Function]}
      >
        text
      </button>
    </Button>
  </div>
</Chip>
`;

exports[`<Chip /> should render a without removable icon 1`] = `
<Chip
  invert={false}
  onClick={[Function]}
  onRemove={[Function]}
  removable={false}
>
  <div
    className="css-9h5w63"
  >
    <Button
      className="css-1al9zi1"
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="css-1al9zi1 css-1rn65c0"
        data-test-id="Button"
        disabled={false}
        onClick={[Function]}
      >
        text
      </button>
    </Button>
  </div>
</Chip>
`;
